File: web/src/feedback.ts
import type { Item } from "./types";

export type FeedbackAction = "signal" | "keep" | "unkeep";
export type FeedbackValue = -1 | 0 | 1;
export interface FeedbackState {
  kept: boolean;
  value: FeedbackValue;
  source?: string;
}

export function applyFeedback(
  state: FeedbackState,
  action: FeedbackAction,
  value: FeedbackValue = 0,
) {
  if (action === "unkeep")
    return { kept: false, value: 0 as const, source: "", preserve: false };
  if (action === "keep")
    return state.value > 0
      ? { ...state, kept: true, source: state.source ?? "", preserve: true }
      : { kept: true, value: 1 as const, source: "heart", preserve: false };
  if (state.kept && value === -1) return undefined;
  return {
    kept: state.kept,
    value: state.kept && value === 0 ? (1 as const) : value,
    source: state.kept && value === 0 ? "heart" : "",
    preserve: false,
  };
}

export function feedbackEligibility(
  item: Pick<Item, "hearted" | "archived">,
  archive = false,
) {
  const live = !archive && !item.archived;
  return {
    signal: live,
    bury: live && !item.hearted,
    lifetime: live && !item.hearted,
  };
}

export function feedbackPatch(
  item: Item,
  action: FeedbackAction,
  value: FeedbackValue = 0,
  archive = false,
) {
  if (action === "signal" && !feedbackEligibility(item, archive).signal)
    return undefined;
  const plan = applyFeedback(
    { kept: item.hearted, value: item.signal },
    action,
    value,
  );
  return plan && { hearted: plan.kept, signal: plan.value };
}

File: internal/feedback/feedback.go
// Package feedback owns the Keep, Boost and Bury rules. Storage executes the
// returned plan, preserving its conditional writes and retry semantics.
package feedback

import (
	"errors"

	"github.com/nuntz/sema/internal/domain"
)

var ErrKept = errors.New("kept items cannot be buried")
var ErrValue = errors.New("value must be -1, 0, or 1")

type Action string

const (
	Signal Action = "signal"
	Keep   Action = "keep"
	Unkeep Action = "unkeep"
)

type State struct {
	Kept   bool
	Value  int
	Source string
}
type Plan struct {
	Kept     bool
	Value    int
	Source   string
	Preserve bool
}

// DeletesSignal reports whether executing the plan removes stored feedback.
func (p Plan) DeletesSignal() bool { return p.Value == 0 && !p.Preserve }

func Kept(item domain.Item) bool {
	return item.ArchiveSK != "" || item.HeartedTS != "" || item.Hearted || domain.IsArchive(item)
}

func Apply(state State, action Action, value int) (Plan, error) {
	switch action {
	case Keep:
		if state.Value > 0 {
			return Plan{Kept: true, Value: state.Value, Source: state.Source, Preserve: true}, nil
		}
		return Plan{Kept: true, Value: 1, Source: "heart"}, nil
	case Unkeep:
		return Plan{}, nil
	case Signal:
		if value < -1 || value > 1 {
			return Plan{}, ErrValue
		}
		if state.Kept && value == -1 {
			return Plan{}, ErrKept
		}
		if state.Kept && value == 0 {
			return Plan{Kept: true, Value: 1, Source: "heart"}, nil
		}
		return Plan{Kept: state.Kept, Value: value}, nil
	default:
		return Plan{}, errors.New("unknown feedback action")
	}
}
